jest/jest-tests.ts

```diff
@@ -367,3 +425,53 @@ describe('strictNullChecks', function () {
         done();
     })
 });
+
+class TestApi {
+    constructor() { };
+    testProp: boolean;
+    testMethod(a: number): string { return ""; }
+}
+
+declare function mockedFunc(a: number): string;
+
+describe('Mocked type', function () {
+    it('Works', function () {
+        const mock: jest.Mocked<TestApi> = new TestApi() as any;
+        mock.testProp;
+        mock.testMethod.mockImplementation(() => 'test');
+        mock.testMethod(5).toUpperCase();
+    });
+});
+
+describe('Mocks', function () {
+    it('jest.fn() without args is a function type', function () {
+        const test = jest.fn();
+        test();
+        new test();
+        test.mock.instances[0];
+        test.mockImplementation(() => { });
+    });
+
+    it('jest.fn() with returned object infers type', function () {
+        const testMock = jest.fn(() => ({ a: 5, test: jest.fn() }));
+
+        testMock(5, 5, 'a');
+        testMock.mockImplementation(() => { });
+        testMock.caller;
+
+        const ins = new testMock();
```

Comment:
The more I think about it it's a little odd for the return type here only to work for constructors. I'd wager there may be more tests that call mock instances as functions than use them as constructors. Should we augment Mock<T>'s call signature to return T to match the construct signature?